File: iOS/APIExample-OC/APIExample-OC/Examples/Advanced/MediaPlayer/MediaPlayer.m
//
//  JoinChannelVideo.m
//  APIExample
//
//  Created by zhaoyongqiang on 2023/7/11.
//

#import "MediaPlayer.h"
#import "KeyCenter.h"
#import <AgoraRtcKit/AgoraRtcKit.h>
#import "VideoView.h"
#import "APIExample_OC-swift.h"
#import "UIView+CSshortFrame.h"

@interface MediaPlayerEntry ()
@property (weak, nonatomic) IBOutlet UITextField *textField;

@end

@implementation MediaPlayerEntry

- (void)viewDidLoad {
    [super viewDidLoad];
}

- (IBAction)onClickJoinButton:(id)sender {
    [self.textField resignFirstResponder];
    
    UIStoryboard *storyBoard = [UIStoryboard storyboardWithName:@"MediaPlayer" bundle:nil];
    BaseViewController *newViewController = [storyBoard instantiateViewControllerWithIdentifier:@"MediaPlayer"];
    newViewController.configs = @{@"channelName": self.textField.text};
    [self.navigationController pushViewController:newViewController animated:YES];
}

@end


@interface MediaPlayer ()<AgoraRtcEngineDelegate, UITextFieldDelegate, AgoraRtcMediaPlayerDelegate>
@property (weak, nonatomic) IBOutlet UIView *containerView;
@property (nonatomic, strong)VideoView *localView;
@property (nonatomic, strong)VideoView *remoteView;
@property (nonatomic, strong)AgoraRtcEngineKit *agoraKit;
@property (nonatomic, weak) IBOutlet UITextField *mediaUrlField;
@property (nonatomic, weak) IBOutlet UIStackView *playerControlStack;
@property (nonatomic, weak) IBOutlet UISlider *playerProgressSlider;
@property (nonatomic, weak) IBOutlet UISlider *playoutVolume;
@property (nonatomic, weak) IBOutlet UISlider *publishVolume;
@property (nonatomic, weak) IBOutlet UILabel *playerDurationLabel;

@property (nonatomic, weak) id<AgoraRtcMediaPlayerProtocol>mediaPlayerKit;
@property (nonatomic, assign) CGFloat originY;
@property (nonatomic, assign) NSUInteger PLAYER_UID;
@property (nonatomic, assign) NSUInteger CAMERA_UID;
@end

@implementation MediaPlayer
- (NSUInteger)PLAYER_UID {
    if (_PLAYER_UID <= 0) {
        _PLAYER_UID = arc4random() % 99999;
    }
    return _PLAYER_UID;
}
- (NSUInteger)CAMERA_UID {
    if (_CAMERA_UID <= 0) {
        _CAMERA_UID = arc4random() % 99999;
    }
    return _CAMERA_UID;
}
- (VideoView *)localView {
    if (_localView == nil) {
        _localView = (VideoView *)[NSBundle loadVideoViewFormType:StreamTypeLocal audioOnly:NO];
    }
    return _localView;
}

- (VideoView *)remoteView {
    if (_remoteView == nil) {
        _remoteView = (VideoView *)[NSBundle loadVideoViewFormType:StreamTypeRemote audioOnly:NO];
    }
    return _remoteView;
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    [textField resignFirstResponder];
    return YES;
}

- (void)keyboardWillAppear: (NSNotification *)noti {
    NSDictionary *keyboardInfo = [noti.userInfo objectForKey:UIKeyboardFrameEndUserInfoKey];
    CGFloat keyboardHeight = ((NSValue *)keyboardInfo).CGRectValue.size.height;
    if (self.originY == 0) {
        self.originY = self.view.centerY_CS;
    }
    CGRect rect = [self.mediaUrlField convertRect:self.mediaUrlField.bounds toView:self.view];
    CGFloat y = self.view.bounds.size.height - rect.origin.y - self.mediaUrlField.bounds.size.height - keyboardHeight;
    if (y < 0) {
        UIViewPropertyAnimator *animator = [[UIViewPropertyAnimator alloc] initWithDuration:0.25 curve:(UIViewAnimationCurveEaseOut) animations:^{
            self.view.centerY_CS = y + self.originY;
        }];
        [animator startAnimation];
    }
}
- (void)keyboardWillDisappear: (NSNotification *)noti {
    UIViewPropertyAnimator *animator = [[UIViewPropertyAnimator alloc] initWithDuration:0.25 curve:(UIViewAnimationCurveEaseOut) animations:^{
        self.view.centerY_CS = self.originY;
    }];
    [animator startAnimation];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.mediaUrlField.delegate = self;
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillAppear:) name:UIKeyboardWillShowNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillDisappear:) name:UIKeyboardWillHideNotification object:nil];
    
    // layout render view
    [self.localView setPlaceholder:@"No Player Loaded".localized];
    [self.remoteView setPlaceholder:@"Remote Host".localized];
    [self.containerView layoutStream:@[self.localView, self.remoteView]];
    
    // set up agora instance when view loaded
    AgoraRtcEngineConfig *config = [[AgoraRtcEngineConfig alloc] init];
    config.appId = KeyCenter.AppId;
    config.channelProfile = AgoraChannelProfileLiveBroadcasting;
    
    self.agoraKit = [AgoraRtcEngineKit sharedEngineWithConfig:config delegate:self];
    
    NSString *channelName = [self.configs objectForKey:@"channelName"];
    // make myself a broadcaster
    [self.agoraKit setClientRole:(AgoraClientRoleBroadcaster)];
    // enable video module and set up video encoding configs
    [self.agoraKit enableAudio];
    [self.agoraKit enableVideo];
    
    AgoraVideoEncoderConfiguration *encoderConfig = [[AgoraVideoEncoderConfiguration alloc] initWithSize:CGSizeMake(960, 540)
                                                                                               frameRate:(AgoraVideoFrameRateFps15)
                                                                                                 bitrate:15
                                                                                         orientationMode:(AgoraVideoOutputOrientationModeFixedPortrait)
                                                                                              mirrorMode:(AgoraVideoMirrorModeAuto)];
    [self.agoraKit setVideoEncoderConfiguration:encoderConfig];
    
    // prepare media player
    self.mediaPlayerKit = [self.agoraKit createMediaPlayerWithDelegate:self];
    [self.mediaPlayerKit setView:self.localView.videoView];
    [self.mediaPlayerKit setRenderMode:(AgoraMediaPlayerRenderModeFit)];
    
    self.playoutVolume.minimumValue = 0;
    self.playoutVolume.maximumValue = 400;
    [self.playoutVolume setValue:[self.mediaPlayerKit getPlayoutVolume]];
    
    self.publishVolume.minimumValue = 0;
    self.publishVolume.maximumValue = 100;
    [self.mediaPlayerKit adjustPublishSignalVolume:50];
    [self.publishVolume setValue:50];
    
    // set up local video to render your local camera preview
    AgoraRtcVideoCanvas *videoCanvas = [[AgoraRtcVideoCanvas alloc] init];
    videoCanvas.uid = 0;
    // the view to be binded
    videoCanvas.view = self.localView.videoView;
    videoCanvas.renderMode = AgoraVideoRenderModeHidden;
    videoCanvas.sourceType = AgoraVideoSourceTypeMediaPlayer;
    videoCanvas.mediaPlayerId = self.mediaPlayerKit.getMediaPlayerId;
    [self.agoraKit setupLocalVideo:videoCanvas];
    // you have to call startPreview to see local video
    [self.agoraKit startPreview];
    
    // Set audio route to speaker
    [self.agoraKit setEnableSpeakerphone:YES];
    
    // start joining channel
    // 1. Users can only see each other after they join the
    // same channel successfully using the same app id.
    // 2. If app certificate is turned on at dashboard, token is needed
    // when joining channel. The channel name and uid used to calculate
    // the token has to match the ones used for channel join
    AgoraRtcChannelMediaOptions *options = [[AgoraRtcChannelMediaOptions alloc] init];
    options.autoSubscribeAudio = YES;
    options.autoSubscribeVideo = YES;
    options.publishCustomAudioTrack = YES;
    options.publishCameraTrack = YES;
    options.publishMicrophoneTrack = YES;
    options.clientRoleType = AgoraClientRoleBroadcaster;
    
    [[NetworkManager shared] generateTokenWithChannelName:channelName uid:self.CAMERA_UID success:^(NSString * _Nullable token) {
        int result = [self.agoraKit joinChannelByToken:token channelId:channelName uid:self.CAMERA_UID mediaOptions:options joinSuccess:nil];
        if (result != 0) {
            // Usually happens with invalid parameters
            // Error code description can be found at:
            // en: https://api-ref.agora.io/en/video-sdk/ios/4.x/documentation/agorartckit/agoraerrorcode
            // cn: https://doc.shengwang.cn/api-ref/rtc/ios/error-code
            NSLog(@"joinChannel call failed: %d, please check your params", result);
        }
    }];
    [self.agoraKit muteRemoteAudioStream:self.PLAYER_UID mute:YES];
    
    
    AgoraRtcChannelMediaOptions *options1 = [[AgoraRtcChannelMediaOptions alloc] init];
    options1.autoSubscribeAudio = NO;
    options1.autoSubscribeVideo = NO;
    options1.clientRoleType = AgoraClientRoleBroadcaster;
    options1.enableAudioRecordingOrPlayout = NO;
    
    AgoraRtcConnection *connection = [[AgoraRtcConnection alloc] initWithChannelId:channelName localUid:self.PLAYER_UID];
    [[NetworkManager shared] generateTokenWithChannelName:channelName uid:self.PLAYER_UID success:^(NSString * _Nullable token) {
        int result = [self.agoraKit joinChannelExByToken:token connection:connection delegate:self mediaOptions:options1 joinSuccess:nil];
        if (result != 0) {
            // Usually happens with invalid parameters
            // Error code description can be found at:
            // en: https://api-ref.agora.io/en/video-sdk/ios/4.x/documentation/agorartckit/agoraerrorcode
            // cn: https://doc.shengwang.cn/api-ref/rtc/ios/error-code
            NSLog(@"joinChannel call failed: %d, please check your params", result);
        }
        [self doOpenMediaUrlWithSender: nil];
    }];
}

- (IBAction)doOpenMediaUrlWithSender:(UIButton *)sender {
    [self.mediaUrlField resignFirstResponder];
    [self.mediaPlayerKit open:self.mediaUrlField.text startPos:0];
}
- (IBAction)doPlayWithSender:(UIButton *)sender {
    [self.mediaPlayerKit play];
}
- (IBAction)doStopWithSender:(UIButton *)sender {
    [self.mediaPlayerKit stop];
}
- (IBAction)doPauseWithSender:(UIButton *)sender {
    [self.mediaPlayerKit pause];
}
- (IBAction)doPublishWithSender:(UIButton *)sender {
    NSString *channelName = [self.configs objectForKey:@"channelName"];
    AgoraRtcChannelMediaOptions *options = [[AgoraRtcChannelMediaOptions alloc] init];
    options.publishMediaPlayerVideoTrack = YES;
    options.publishMediaPlayerAudioTrack = YES;
    options.publishMediaPlayerId = self.mediaPlayerKit.getMediaPlayerId;
    options.publishCameraTrack = NO;
    options.clientRoleType = AgoraClientRoleBroadcaster;
    
    AgoraRtcConnection *connection = [[AgoraRtcConnection alloc] initWithChannelId:channelName localUid:self.PLAYER_UID];
    [self.agoraKit updateChannelExWithMediaOptions:options connection:connection];
}
- (IBAction)dloStopPublish:(UIButton *)sender {
    NSString *channelName = [self.configs objectForKey:@"channelName"];
    AgoraRtcChannelMediaOptions *options = [[AgoraRtcChannelMediaOptions alloc] init];
    options.publishMediaPlayerVideoTrack = NO;
    options.publishMediaPlayerAudioTrack = NO;
    options.publishMediaPlayerId = self.mediaPlayerKit.getMediaPlayerId;
    options.publishCameraTrack = NO;
    options.clientRoleType = AgoraClientRoleBroadcaster;
    
    AgoraRtcConnection *connection = [[AgoraRtcConnection alloc] initWithChannelId:channelName localUid:self.PLAYER_UID];
    [self.agoraKit updateChannelExWithMediaOptions:options connection:connection];
}
- (IBAction)doAdjustPlayoutVolumeWithSender:(UISlider *)sender {
    [self.mediaPlayerKit adjustPlayoutVolume:sender.value];
}
- (IBAction)doAdjustPublishVolumeWithSender:(UISlider *)sender {
    [self.mediaPlayerKit adjustPublishSignalVolume:sender.value];
}

- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    [self.agoraKit disableAudio];
    [self.agoraKit disableVideo];
    [self.agoraKit leaveChannel:nil];
    [self.agoraKit destroyMediaPlayer:self.mediaPlayerKit];
    [AgoraRtcEngineKit destroy];
}

- (void)AgoraRtcMediaPlayer:(id<AgoraRtcMediaPlayerProtocol>)playerKit didChangedToState:(AgoraMediaPlayerState)state reason:(AgoraMediaPlayerReason)reason {
    [LogUtil log:[NSString stringWithFormat:@"player rtc channel publish helper state changed to: %ld error: %ld", state, reason]];
    __weak MediaPlayer *weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        switch (state) {
            case AgoraMediaPlayerStateFailed:
                [weakSelf showAlertWithTitle:[NSString stringWithFormat:@"media player error: %ld", reason]];
                break;
            
            case AgoraMediaPlayerStateOpenCompleted:
            {
                NSInteger duation = playerKit.getDuration;
                [weakSelf.playerControlStack setHidden:NO];
                weakSelf.playerDurationLabel.text = [NSString stringWithFormat:@"%02ld : %02ld", duation / 60000, duation % 60000 / 1000];
                [weakSelf.playerProgressSlider setValue:0 animated:YES];
            }
                break;
                
            case AgoraMediaPlayerStateStopped:
            {
                [weakSelf.playerControlStack setHidden:YES];
                [weakSelf.playerProgressSlider setValue:0 animated:YES];
                weakSelf.playerDurationLabel.text = @"00 : 00";
                NSString *channelName = [self.configs objectForKey:@"channelName"];
                AgoraRtcChannelMediaOptions *option = [[AgoraRtcChannelMediaOptions alloc] init];
                option.publishMediaPlayerVideoTrack = NO;
                option.publishMediaPlayerAudioTrack = NO;
                AgoraRtcConnection *connection = [[AgoraRtcConnection alloc] initWithChannelId:channelName localUid:self.PLAYER_UID];
                [weakSelf.agoraKit updateChannelExWithMediaOptions:option connection:connection];
            }
            default:
                break;
        }
    });
}

- (void)AgoraRtcMediaPlayer:(id<AgoraRtcMediaPlayerProtocol>)playerKit didChangedToPosition:(NSInteger)position {
    NSInteger duration = self.mediaPlayerKit.getDuration;
    float progress = 0;
    long left = 0;
    if (duration > 0) {
        progress = position / duration;
        left = (self.mediaPlayerKit.getDuration - position) / 1000;
    }
    dispatch_async(dispatch_get_main_queue(), ^{
        self.playerDurationLabel.text = [NSString stringWithFormat:@"%02ld : %02ld", left/60, left % 60];
        if (self.playerProgressSlider.isTouchInside) {
            [self.playerProgressSlider setValue:progress animated:YES];
        }
    });
}


/// callback when error occured for agora sdk, you are recommended to display the error descriptions on demand
/// to let user know something wrong is happening
/// Error code description can be found at:
/// en: https://api-ref.agora.io/en/video-sdk/ios/4.x/documentation/agorartckit/agoraerrorcode
/// cn: https://doc.shengwang.cn/api-ref/rtc/ios/error-code
/// @param errorCode error code of the problem
- (void)rtcEngine:(AgoraRtcEngineKit *)engine didOccurError:(AgoraErrorCode)errorCode {
    [LogUtil log:[NSString stringWithFormat:@"Error %ld occur",errorCode] level:(LogLevelError)];
}

- (void)rtcEngine:(AgoraRtcEngineKit *)engine didJoinChannel:(NSString *)channel withUid:(NSUInteger)uid elapsed:(NSInteger)elapsed {
    [LogUtil log:[NSString stringWithFormat:@"Join %@ with uid %lu elapsed %ldms", channel, uid, elapsed] level:(LogLevelDebug)];
    self.localView.uid = uid;
}

/// callback when a remote user is joinning the channel, note audience in live broadcast mode will NOT trigger this event
/// @param uid uid of remote joined user
/// @param elapsed time elapse since current sdk instance join the channel in ms
- (void)rtcEngine:(AgoraRtcEngineKit *)engine didJoinedOfUid:(NSUInteger)uid elapsed:(NSInteger)elapsed {
    [LogUtil log:[NSString stringWithFormat:@"remote user join: %lu %ldms", uid, elapsed] level:(LogLevelDebug)];
    // Only one remote video view is available for this
    // tutorial. Here we check if there exists a surface
    // view tagged as this uid.
    AgoraRtcVideoCanvas *videoCanvas = [[AgoraRtcVideoCanvas alloc]init];
    videoCanvas.uid = uid;
    // the view to be binded
    videoCanvas.view = self.remoteView.videoView;
    videoCanvas.renderMode = AgoraVideoRenderModeHidden;
    [self.agoraKit setupRemoteVideo:videoCanvas];
    self.remoteView.uid = uid;
}

/// callback when a remote user is leaving the channel, note audience in live broadcast mode will NOT trigger this event
/// @param uid uid of remote joined user
/// @param reason reason why this user left, note this event may be triggered when the remote user
/// become an audience in live broadcasting profile
- (void)rtcEngine:(AgoraRtcEngineKit *)engine didOfflineOfUid:(NSUInteger)uid reason:(AgoraUserOfflineReason)reason {
    // to unlink your view from sdk, so that your view reference will be released
    // note the video will stay at its last frame, to completely remove it
    // you will need to remove the EAGL sublayer from your binded view
    AgoraRtcVideoCanvas *videoCanvas = [[AgoraRtcVideoCanvas alloc]init];
    videoCanvas.uid = uid;
    // the view to be binded
    videoCanvas.view = nil;
    [self.agoraKit setupRemoteVideo:videoCanvas];
    self.remoteView.uid = 0;
    [LogUtil log:[NSString stringWithFormat:@"remote user left: %lu", uid] level:(LogLevelDebug)];
}

@end

File: iOS/APIExample/APIExample/Examples/Advanced/MediaPlayer/MediaPlayer.swift
//
//  MediaPlayer.swift
//  APIExample
//
//  Created by 张乾泽 on 2020/4/17.
//  Copyright © 2020 Agora Corp. All rights reserved.
//
import UIKit
import AGEVideoLayout
import AgoraRtcKit

let CAMERA_UID = UInt.random(in: 1001...2000)
let PLAYER_UID = UInt.random(in: 2001...3000)

class MediaPlayerEntry : UIViewController
{
    @IBOutlet weak var joinButton: UIButton!
    @IBOutlet weak var channelTextField: UITextField!
    let identifier = "MediaPlayer"
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    @IBAction func doJoinPressed(sender: UIButton) {
        guard let channelName = channelTextField.text else {return}
        //resign channel text field
        channelTextField.resignFirstResponder()
        
        let storyBoard: UIStoryboard = UIStoryboard(name: identifier, bundle: nil)
        // create new view controller every time to ensure we get a clean vc
        guard let newViewController = storyBoard.instantiateViewController(withIdentifier: identifier) as? BaseViewController else {return}
        newViewController.title = channelName
        newViewController.configs = ["channelName":channelName]
        navigationController?.pushViewController(newViewController, animated: true)
    }
    
}

class MediaPlayerMain: BaseViewController, UITextFieldDelegate {
    var localVideo = Bundle.loadView(fromNib: "VideoView", withType: VideoView.self)
    var remoteVideo = Bundle.loadView(fromNib: "VideoView", withType: VideoView.self)
    
    @IBOutlet weak var container: AGEVideoContainer!
    @IBOutlet weak var mediaUrlField: UITextField!
    @IBOutlet weak var playerControlStack: UIStackView!
    @IBOutlet weak var playerProgressSlider: UISlider!
    @IBOutlet weak var playoutVolume: UISlider!
    @IBOutlet weak var publishVolume: UISlider!
    @IBOutlet weak var playerDurationLabel: UILabel!
    
    var agoraKit: AgoraRtcEngineKit!
    var mediaPlayerKit: AgoraRtcMediaPlayerProtocol!
    
    private var originY: CGFloat = 0
    
    // indicate if current instance has joined channel
    var isJoined: Bool = false
    
    @objc func keyboardWillAppear(notification: NSNotification) {
        let keyboardinfo = notification.userInfo![UIResponder.keyboardFrameEndUserInfoKey]
        let keyboardheight:CGFloat = (keyboardinfo as AnyObject).cgRectValue.size.height
   
        if self.originY == 0 {
            self.originY = self.view.centerY_CS
        }
        let rect = self.mediaUrlField.convert(self.mediaUrlField.bounds, to: self.view)
        let y = self.view.bounds.height - rect.origin.y - self.mediaUrlField.bounds.height - keyboardheight

        if y < 0 {
            let animator = UIViewPropertyAnimator(duration: 0.2, curve: .easeOut) {
                self.view.centerY_CS = y + self.originY
            }
            animator.startAnimation()
        }
    }
    
    @objc func keyboardWillDisappear(notification:NSNotification){
        let animator = UIViewPropertyAnimator(duration: 0.2, curve: .easeOut) {
            self.view.centerY_CS = self.originY
            self.originY = 0
        }
        animator.startAnimation()
    }
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        return true
    }
    
    func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
        if text == "\n" {
            textView.resignFirstResponder()
            return true
        }
        return false
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        mediaUrlField.delegate = self
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillAppear), name: UIResponder.keyboardWillShowNotification, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillDisappear), name: UIResponder.keyboardWillHideNotification, object: nil)
        // layout render view
        localVideo.setPlaceholder(text: "No Player Loaded")
        remoteVideo.setPlaceholder(text: "Remote Host".localized)
        container.layoutStream1x2(views: [localVideo, remoteVideo])
        
        // set up agora instance when view loaded
        let config = AgoraRtcEngineConfig()
        config.appId = KeyCenter.AppId
        config.areaCode = GlobalSettings.shared.area
        config.channelProfile = .liveBroadcasting
        agoraKit = AgoraRtcEngineKit.sharedEngine(with: config, delegate: self)
        // Configuring Privatization Parameters
        Util.configPrivatization(agoraKit: agoraKit)
        agoraKit.setLogFile(LogUtils.sdkLogPath())
        
        // enable video module and set up video encoding configs
        agoraKit.enableVideo()
        agoraKit.enableAudio()
        let resolution = (GlobalSettings.shared.getSetting(key: "resolution")?.selectedOption().value as? CGSize) ?? .zero
        let fps = (GlobalSettings.shared.getSetting(key: "fps")?.selectedOption().value as? AgoraVideoFrameRate) ?? .fps15
        let orientation = (GlobalSettings.shared.getSetting(key: "orientation")?.selectedOption().value as? AgoraVideoOutputOrientationMode) ?? .fixedPortrait
        agoraKit.setVideoEncoderConfiguration(AgoraVideoEncoderConfiguration(size: resolution,
                                                                             frameRate: fps,
                                                                             bitrate: AgoraVideoBitrateStandard,
                                                                             orientationMode: orientation,
                                                                             mirrorMode: .auto))
        // get channel name from configs
        guard let channelName = configs["channelName"] as? String else { return }
                
        // prepare media player
        mediaPlayerKit = agoraKit.createMediaPlayer(with: self)
        mediaPlayerKit.setView(localVideo.videoView)
        mediaPlayerKit.setRenderMode(.fit)
        
        playoutVolume.minimumValue = 0
        playoutVolume.maximumValue = 400
        playoutVolume.integerValue = Int(mediaPlayerKit.getPlayoutVolume())
        
        publishVolume.minimumValue = 0
        publishVolume.maximumValue = 100
        mediaPlayerKit.adjustPublishSignalVolume(50)
        publishVolume.integerValue = 50
        
        let videoCanvas = AgoraRtcVideoCanvas()
        videoCanvas.view = localVideo.videoView
        videoCanvas.renderMode = .hidden
        videoCanvas.sourceType = .mediaPlayer
        videoCanvas.mediaPlayerId = mediaPlayerKit.getMediaPlayerId()
        agoraKit.setupLocalVideo(videoCanvas)
        let option = AgoraRtcChannelMediaOptions()
        option.publishCameraTrack = true
        option.publishCustomAudioTrack = true
        option.publishMicrophoneTrack = true
        option.autoSubscribeAudio = true
        option.autoSubscribeVideo = true
        option.clientRoleType = GlobalSettings.shared.getUserRole()
        NetworkManager.shared.generateToken(channelName: channelName, uid: CAMERA_UID, success: { token in
            self.agoraKit.joinChannel(byToken: token, channelId: channelName, uid: CAMERA_UID, mediaOptions: option, joinSuccess: nil)
        })
        agoraKit.muteRemoteAudioStream(PLAYER_UID, mute: true)

        let option1 = AgoraRtcChannelMediaOptions()
        option1.autoSubscribeAudio = false
        option1.autoSubscribeVideo = false
        option1.clientRoleType = GlobalSettings.shared.getUserRole()
        option1.enableAudioRecordingOrPlayout = false
        let connection = AgoraRtcConnection()
        connection.channelId = channelName
        connection.localUid = PLAYER_UID
        NetworkManager.shared.generateToken(channelName: channelName, uid: PLAYER_UID) { token in
            let result1 = self.agoraKit.joinChannelEx(byToken: token, connection: connection, delegate: self, mediaOptions: option1, joinSuccess: nil)
            if result1 != 0 {
                // Usually happens with invalid parameters
                // Error code description can be found at:
                // en: https://api-ref.agora.io/en/voice-sdk/macos/3.x/Constants/AgoraErrorCode.html#content
                // cn: https://docs.agora.io/cn/Voice/API%20Reference/oc/Constants/AgoraErrorCode.html
                self.showAlert(title: "Error", message: "joinChannel call failed, please check your params")
            }
            self.doOpenMediaUrl(sender: UIButton())
        }
    }
    
    @IBAction func doOpenMediaUrl(sender: UIButton) {
        guard let url = mediaUrlField.text else { return }
        //resign text field
        mediaUrlField.resignFirstResponder()
        mediaPlayerKit.open(url, startPos: 0)
    }
    
    @IBAction func doPlay(sender: UIButton) {
        mediaPlayerKit.play()// get channel name from configs
    }
    
    @IBAction func doStop(sender: UIButton) {
        mediaPlayerKit.stop()
    }
    
    @IBAction func doPause(sender: UIButton) {
        mediaPlayerKit.pause()
    }
    
    @IBAction func doPublish(sender: UIButton) {
        guard let channelName = configs["channelName"] as? String else { return }
        let option = AgoraRtcChannelMediaOptions()
        option.publishMediaPlayerVideoTrack = true
        option.publishMediaPlayerAudioTrack = true
        option.publishMediaPlayerId = Int(mediaPlayerKit.getMediaPlayerId())
        option.publishCameraTrack = false
        let connection = AgoraRtcConnection()
        connection.channelId = channelName
        connection.localUid = PLAYER_UID
        agoraKit.updateChannelEx(with: option, connection: connection)

    }
    @IBAction func dloStopPublish(_ sender: Any) {
        guard let channelName = configs["channelName"] as? String else { return }
        let option = AgoraRtcChannelMediaOptions()
        option.publishMediaPlayerVideoTrack = false
        option.publishMediaPlayerAudioTrack = false
        option.publishMediaPlayerId = Int(mediaPlayerKit.getMediaPlayerId())
        option.publishCameraTrack = false
        let connection = AgoraRtcConnection()
        connection.channelId = channelName
        connection.localUid = PLAYER_UID
        agoraKit.updateChannelEx(with: option, connection: connection)
    }
    
    @IBAction func doAdjustPlayoutVolume(sender: UISlider) {
        mediaPlayerKit.adjustPlayoutVolume(Int32(Int(sender.value)))
    }
    
    @IBAction func doAdjustPublishVolume(sender: UISlider) {
        mediaPlayerKit.adjustPublishSignalVolume(Int32(Int(sender.value)))
    }
    
    override func willMove(toParent parent: UIViewController?) {
        if parent == nil {
            // leave channel when exiting the view
            if isJoined {
                agoraKit.disableAudio()
                agoraKit.disableVideo()
                agoraKit.leaveChannel { (stats) -> Void in
                    LogUtils.log(message: "left channel, duration: \(stats.duration)", level: .info)
                }
            }
            agoraKit.destroyMediaPlayer(mediaPlayerKit)
            mediaPlayerKit = nil
            AgoraRtcEngineKit.destroy()
        }
    }
}

/// agora rtc engine delegate events
extension MediaPlayerMain: AgoraRtcEngineDelegate {
    /// callback when warning occured for agora sdk, warning can usually be ignored, still it's nice to check out
    /// what is happening
    /// Warning code description can be found at:
    /// en: https://api-ref.agora.io/en/voice-sdk/ios/3.x/Constants/AgoraWarningCode.html
    /// cn: https://docs.agora.io/cn/Voice/API%20Reference/oc/Constants/AgoraWarningCode.html
    /// @param warningCode warning code of the problem
    func rtcEngine(_ engine: AgoraRtcEngineKit, didOccurWarning warningCode: AgoraWarningCode) {
        LogUtils.log(message: "warning: \(warningCode.description)", level: .warning)
    }
    
    /// callback when error occured for agora sdk, you are recommended to display the error descriptions on demand
    /// to let user know something wrong is happening
    /// Error code description can be found at:
    /// en: https://api-ref.agora.io/en/voice-sdk/macos/3.x/Constants/AgoraErrorCode.html#content
    /// cn: https://docs.agora.io/cn/Voice/API%20Reference/oc/Constants/AgoraErrorCode.html
    /// @param errorCode error code of the problem
    func rtcEngine(_ engine: AgoraRtcEngineKit, didOccurError errorCode: AgoraErrorCode) {
        LogUtils.log(message: "error: \(errorCode)", level: .error)
        self.showAlert(title: "Error", message: "Error \(errorCode.description) occur")
    }
    
    func rtcEngine(_ engine: AgoraRtcEngineKit, didJoinChannel channel: String, withUid uid: UInt, elapsed: Int) {
        isJoined = true
        LogUtils.log(message: "local user join: \(uid) \(elapsed)ms", level: .info)
    }
    
    /// callback when a remote user is joinning the channel, note audience in live broadcast mode will NOT trigger this event
    /// @param uid uid of remote joined user
    /// @param elapsed time elapse since current sdk instance join the channel in ms
    func rtcEngine(_ engine: AgoraRtcEngineKit, didJoinedOfUid uid: UInt, elapsed: Int) {
        if uid == CAMERA_UID || uid == PLAYER_UID {
            return
        }
        LogUtils.log(message: "remote user join: \(uid) \(elapsed)ms", level: .info)
        
        // Only one remote video view is available for this
        // tutorial. Here we check if there exists a surface
        // view tagged as this uid.
        let videoCanvas = AgoraRtcVideoCanvas()
        videoCanvas.uid = uid
        // the view to be binded
        videoCanvas.view = remoteVideo.videoView
        videoCanvas.renderMode = .hidden
        agoraKit.setupRemoteVideo(videoCanvas)
    }
    
    /// callback when a remote user is leaving the channel, note audience in live broadcast mode will NOT trigger this event
    /// @param uid uid of remote joined user
    /// @param reason reason why this user left, note this event may be triggered when the remote user
    /// become an audience in live broadcasting profile
    func rtcEngine(_ engine: AgoraRtcEngineKit, didOfflineOfUid uid: UInt, reason: AgoraUserOfflineReason) {
        LogUtils.log(message: "remote user left: \(uid) reason \(reason)", level: .info)
        
        // to unlink your view from sdk, so that your view reference will be released
        // note the video will stay at its last frame, to completely remove it
        // you will need to remove the EAGL sublayer from your binded view
        let videoCanvas = AgoraRtcVideoCanvas()
        videoCanvas.uid = uid
        // the view to be binded
        videoCanvas.view = nil
        videoCanvas.renderMode = .hidden
        agoraKit.setupRemoteVideo(videoCanvas)
    }
}

extension MediaPlayerMain: AgoraRtcMediaPlayerDelegate {
    func AgoraRtcMediaPlayer(_ playerKit: AgoraRtcMediaPlayerProtocol, didChangedTo state: AgoraMediaPlayerState, error: AgoraMediaPlayerError) {
        LogUtils.log(message: "player rtc channel publish helper state changed to: \(state.rawValue), error: \(error.rawValue)", level: .info)
        DispatchQueue.main.async {[weak self] in
            guard let weakself = self else { return }
            switch state.rawValue {
            case 100: // failed
                weakself.showAlert(message: "media player error: \(error.rawValue)")
                break
            case 2: // openCompleted
                let duration = weakself.mediaPlayerKit.getDuration()
                weakself.playerControlStack.isHidden = false
                weakself.playerDurationLabel.text = "\(String(format: "%02d", duration / 60000)) : \(String(format: "%02d", duration % 60000 / 1000))"
                weakself.playerProgressSlider.setValue(0, animated: true)
                break
            case 7: // stopped
                weakself.playerControlStack.isHidden = true
                weakself.playerProgressSlider.setValue(0, animated: true)
                weakself.playerDurationLabel.text = "00 : 00"
                guard let channelName = weakself.configs["channelName"] as? String else { return }
                let option = AgoraRtcChannelMediaOptions()
                option.publishMediaPlayerVideoTrack = false
                option.publishMediaPlayerAudioTrack = false
                let connection = AgoraRtcConnection()
                connection.channelId = channelName
                connection.localUid = PLAYER_UID
                weakself.agoraKit.updateChannelEx(with: option, connection: connection)
                break
            default: break
            }
        }
    }
    
    func AgoraRtcMediaPlayer(_ playerKit: AgoraRtcMediaPlayerProtocol, didChangedTo position: Int) {
        let duration = Float(mediaPlayerKit.getDuration())
        var progress: Float = 0
        var left: Int = 0
        if duration > 0 {
            progress = Float(position) / duration
            left = Int((mediaPlayerKit.getDuration() - position)) / 1000
        }
        DispatchQueue.main.async {[weak self] in
            guard let weakself = self else { return }
            weakself.playerDurationLabel.text = "\(String(format: "%02d", left / 60)) : \(String(format: "%02d", left % 60))"
            if !weakself.playerProgressSlider.isTouchInside {
                weakself.playerProgressSlider.setValue(progress, animated: true)
            }
        }
    }
}
